testng-core/src/main/java/org/testng/internal/Parameters.java

```diff
@@ -495,6 +495,16 @@ private static IDataProviderMethod findDataProvider(
     if (dp != null) {
       String dataProviderName = dp.getDataProvider();
       Class<?> dataProviderClass = dp.getDataProviderClass();
+      boolean isDynamicDataProvider = dataProviderClass == null
+          && !dp.getDataProviderDynamicClass().isEmpty();
+      if (isDynamicDataProvider) {
+        try {
+          dataProviderClass = new DataProviderLoader()
```

Comment:
We are creating on the fly instances of `DataProviderLoader()` to load the class that was specified as a string attribute in the `@Test` annotation. This means that this object of `DataProviderLoader()` immediately becomes eligible for a GC. So if the loader object gets GC'ed then all the classes that it caused to load would be unloaded as well. 

So what is going to be the expected behaviour when the `DataProviderLoader` object got GC'ed even before a data provider was actually invoked? Wouldn't that cause `ClassNotFoundException` or `NoClassDefFoundError` for the test case which is using a dynamic data provider ?

Reply:
Class holds a reference to it's classloader. So, in order to garbage collect a classloader, all class instances should be collected first.

Reply:
That's right.